### ISSUE TITLE
Add concat methods to Tuple classes

### ISSUE THREAD

**Issue by jlorenzen** [vavr-io/vavr#2215]
It would be handy if the individual Tuple classes add concat or append methods to them. I've ran into two instances lately where they would have been helpful. The jOOL [Tuple](http://www.jooq.org/products/jOO%CE%BB/javadoc/0.9.5/org/jooq/lambda/tuple/Tuple2.html) classes have them. I would use them but the jOOQ Tuple classes don't have the `apple` method which I've come to love in the vavr Tuple classes. And I'm trying to standardize on using just vavr.

**Comment by nfekete** [comment]
@jlorenzen do you mean _widening_ and _narrowing_ `Tuple`s from the right?

**Comment by jlorenzen** [comment]
@nfekete Yeah I think so. Basically the following:
```java
final Tuple2<String, String> user = Tuple.of("John", "Doe");
final Tuple3<String, String, String> userWithEmail = user.concat("john.doe@example.com");
```

**Comment by jlorenzen** [comment]
@nfekete @danieldietrich I would be interested in making these changes if you guys would welcome them and offer any suggestions?

**Comment by ruslansennov** [comment]
@jlorenzen The `append` and `prepend` methods were added and removed twice. It was a difficult decision, see the [last discussion](https://github.com/vavr-io/vavr/pull/1559#discussion_r80514299).

**Comment by jlorenzen** [comment]
@ruslansennov That last discussion didn't really go deep into why it was removed. There is hint of people not liking that `Tuple8` (the last numbered Tuple) had to throw an exception since there isn't a `Tuple9` class.
Looking at the last jOOL Tuple class, Tuple16, they just don't add a `concat` method. That would seem like a reasonable compromise.
Here is the reason I'd like to see the vavr Tuple classes have at least a `concat` or `append` method. Lately I've been using the `Validation` class to perform validation on HTTP request parameters. I then convert those to a Tuple to be used in the next step. So something like the following:
```java
private SignInResult signIn(final String email, final String password) {
    validate(email, password)
        .flatMap(emailAndPassword -> emailAndPassword.apply(signInService::signIn));
}

private Either<Throwable, Tuple2<Email, Password> validate(final String email, final String password) {
    // create validation objects for each param
    return Validation.combine(emailValidation, passwordValidation)
        .ap(Tuple::of)
        .toEither();
}
```

Given my example above, my problem is the `signInService.signIn` method takes a third parameter that doesn't come from client input. Perhaps it might include the IP Address of the client. So here is what I would like to do:
```java
private SignInResult signIn(final String email, final String password) {
    validate(email, password)
        .map(emailAndPassword -> emailAndPassword.append(getClientAddress()))
        .flatMap(signInValues -> signInValues.apply(signInService::signIn));
}
```

Instead I have to reference the tuple values (which I can't stand to do because I feel it hurts readability):
```java
private SignInResult signIn(final String email, final String password) {
    validate(email, password)
        .flatMap(emailAndPassword -> signInService.signIn(emailAndPassword._1, emailAndPassword._2, getClientAddress()));
}
```

I really enjoy using the `Tuple.apply` methods because it helps me avoid referencing the individual values using `._n`.

**Comment by danieldietrich** [comment]
@jlorenzen Yes, I understand the benefits of having it. So, let's add the `concat` method.

(Having an `append` might ignite the desire to also add a `prepend`, which is unnecessary when swapping the tuples and using `concat` instead.)

**Comment by jlorenzen** [comment]
@danieldietrich Hey I took an initial stab at making these changes and I would appreciate some feedback before I go any further. See the PR here: https://github.com/vavr-io/vavr/pull/2221. Thanks!

**Comment by danieldietrich** [comment]
Fixed with #2221
